core/src/main/scala/org/http4s/multipart/MultipartParser.scala

```diff
@@ -0,0 +1,166 @@
+package org.http4s
+package multipart
+
+import scala.language.postfixOps
+
+import scodec.bits.ByteVector
+
+import scalaz.{-\/, \/-, \/}
+import scalaz.stream.{process1, Process, Process1, Writer1}
+
+/** A low-level multipart-parsing pipe.  Most end users will prefer EntityDecoder[Multipart]. */
+object MultipartParser {
+  import Process._
+
+  private val logger = org.log4s.getLogger
+
+  private val CRLF = ByteVector('\r', '\n')
+  private val DASHDASH = ByteVector('-', '-')
+
+  def parse(boundary: Boundary): Writer1[Headers, ByteVector, ByteVector] = {
+    val boundaryBytes = boundary.toByteVector
+    val startLine = DASHDASH ++ boundaryBytes
+    val endLine = startLine ++ DASHDASH
+
+    def receiveLine(leading: Option[ByteVector]): Writer1[ByteVector, ByteVector, (ByteVector, Option[ByteVector])] = {
+      def handle(bv: ByteVector): Writer1[ByteVector, ByteVector, (ByteVector, Option[ByteVector])] = {
+        val index = bv indexOfSlice CRLF
+
+        if (index >= 0) {
+          val (head, tailPre) = bv splitAt index
+          val tail = tailPre drop 2       // remove the line feed characters
+          val tailM = Some(tail) filter { !_.isEmpty }
+          emit(\/-((head, tailM)))
+        }
+        else if (bv.nonEmpty && bv.get(bv.length - 1) == '\r') {
+          // The CRLF may have split across chunks.
+          val (head, cr) = bv.splitAt(bv.length - 1)
+          emit(-\/(head)) ++ receive1(next => handle(cr ++ next))
+        }
+        else {
+          emit(-\/(bv)) ++ receiveLine(None)
+        }
+      }
+
+      leading map handle getOrElse receive1(handle)
+    }
+
+    def receiveCollapsedLine(leading: Option[ByteVector]): Process1[ByteVector, (ByteVector, Option[ByteVector])] = {
+      receiveLine(leading) pipe process1.fold((ByteVector.empty, None: Option[ByteVector])) {
+        case ((acc, tail), -\/(left)) => (acc ++ left, tail)
+        case ((acc, _), \/-((term, tail))) => (acc ++ term, tail)
+      }
+    }
+
+    lazy val start: Writer1[Headers, ByteVector, ByteVector] = {
```

Comment:
Also beware: this `lazy` uses the `object MultipartParser` instance as the monitor lock.